### ISSUE TITLE
InternalError: LOCK TABLE can only be used in transaction blocks

### ISSUE THREAD

**Issue by rrix** [matrix-org/synapse#1890]
```
  C-c C-cFeb 07 21:24:51 hypervisor01 docker[28244]: Traceback (most recent call last):
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/federation/federation_server.py", line 208, in received_edu
Feb 07 21:24:51 hypervisor01 docker[28244]: yield self.edu_handlers[edu_type](origin, content)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/internet/defer.py", line 1297, in _inlineCallbacks
Feb 07 21:24:51 hypervisor01 docker[28244]: result = result.throwExceptionIntoGenerator(g)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
Feb 07 21:24:51 hypervisor01 docker[28244]: return g.throw(self.type, self.value, self.tb)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/util/metrics.py", line 58, in measured_func
Feb 07 21:24:51 hypervisor01 docker[28244]: r = yield func(self, *args, **kwargs)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/internet/defer.py", line 1297, in _inlineCallbacks
Feb 07 21:24:51 hypervisor01 docker[28244]: result = result.throwExceptionIntoGenerator(g)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
Feb 07 21:24:51 hypervisor01 docker[28244]: return g.throw(self.type, self.value, self.tb)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/handlers/device.py", line 319, in _incoming_device_list_update
Feb 07 21:24:51 hypervisor01 docker[28244]: user_id, devices, stream_id,
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/internet/defer.py", line 1297, in _inlineCallbacks
Feb 07 21:24:51 hypervisor01 docker[28244]: result = result.throwExceptionIntoGenerator(g)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/python/failure.py", line 389, in throwExceptionIntoGenerator
Feb 07 21:24:51 hypervisor01 docker[28244]: return g.throw(self.type, self.value, self.tb)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 311, in runInteraction
Feb 07 21:24:51 hypervisor01 docker[28244]: inner_func, *args, **kwargs
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/python/threadpool.py", line 246, in inContext
Feb 07 21:24:51 hypervisor01 docker[28244]: result = inContext.theWork()
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/python/threadpool.py", line 262, in <lambda>
Feb 07 21:24:51 hypervisor01 docker[28244]: inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 118, in callWithContext
Feb 07 21:24:51 hypervisor01 docker[28244]: return self.currentContext().callWithContext(ctx, func, *args, **kw)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/python/context.py", line 81, in callWithContext
Feb 07 21:24:51 hypervisor01 docker[28244]: return func(*args,**kw)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/enterprise/adbapi.py", line 307, in _runWithConnection
Feb 07 21:24:51 hypervisor01 docker[28244]: compat.reraise(excValue, excTraceback)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib64/python2.7/site-packages/twisted/enterprise/adbapi.py", line 298, in _runWithConnection
Feb 07 21:24:51 hypervisor01 docker[28244]: result = func(conn, *args, **kw)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 305, in inner_func
Feb 07 21:24:51 hypervisor01 docker[28244]: func, *args, **kwargs
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 234, in _new_transaction
Feb 07 21:24:51 hypervisor01 docker[28244]: r = func(txn, *args, **kwargs)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/devices.py", line 244, in _update_remote_device_list_cache_txn
Feb 07 21:24:51 hypervisor01 docker[28244]: "stream_id": stream_id,
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 477, in _simple_upsert_txn
Feb 07 21:24:51 hypervisor01 docker[28244]: self.database_engine.lock_table(txn, table)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/engines/postgres.py", line 64, in lock_table
Feb 07 21:24:51 hypervisor01 docker[28244]: txn.execute("LOCK TABLE %s in EXCLUSIVE MODE" % (table,))
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 78, in execute
Feb 07 21:24:51 hypervisor01 docker[28244]: self._do_execute(self.txn.execute, sql, *args)
Feb 07 21:24:51 hypervisor01 docker[28244]: File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 102, in _do_execute
Feb 07 21:24:51 hypervisor01 docker[28244]: sql, *args
Feb 07 21:24:51 hypervisor01 docker[28244]: InternalError: LOCK TABLE can only be used in transaction blocks
```

**Comment by rrix** [comment]
```
Feb 08 20:24:08 hypervisor01 docker[29567]: 2017-02-09 04:24:08,955 - synapse.util.metrics - 101 - WARNING -  - Context has unexpectedly changed from 'Measure@7f9364dde600' to 'sentinel'. ('_send_new_transaction')
```

possibly relevant?

**Comment by chazdiezal** [comment]
i'm experiencing the same issue

**Comment by richvdh** [comment]
possibly related to https://github.com/matrix-org/synapse/issues/2135; Shrike is suffering from this one too

**Comment by richvdh** [comment]
This also bites during key requests:

```
Apr 26 18:17:13 iris synapse[7386]: synapse.crypto.keyring: [] Unable to get key from 'matrix.org': InternalError LOCK TABLE can only be used in transaction blocks
                                    Traceback (most recent call last):
                                      File "/usr/lib/python2.7/site-packages/synapse/crypto/keyring.py", line 354, in get_key
                                        server_name_and_key_ids, perspective_name, perspective_keys
                                      File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1299, in _inlineCallbacks
                                        result = result.throwExceptionIntoGenerator(g)
                                      File "/usr/lib/python2.7/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
                                        return g.throw(self.type, self.value, self.tb)
                                      File "/usr/lib/python2.7/site-packages/synapse/crypto/keyring.py", line 482, in get_server_verify_key_v2_indirect
                                        perspective_name, response, only_from_server=False
                                      File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1299, in _inlineCallbacks
                                        result = result.throwExceptionIntoGenerator(g)
                                      File "/usr/lib/python2.7/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
                                        return g.throw(self.type, self.value, self.tb)
                                      File "/usr/lib/python2.7/site-packages/synapse/crypto/keyring.py", line 635, in process_v2_response
                                        )).addErrback(unwrapFirstError)
                                      File "/usr/lib/python2.7/site-packages/twisted/internet/defer.py", line 1299, in _inlineCallbacks
                                        result = result.throwExceptionIntoGenerator(g)
                                      File "/usr/lib/python2.7/site-packages/twisted/python/failure.py", line 393, in throwExceptionIntoGenerator
                                        return g.throw(self.type, self.value, self.tb)
                                      File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 319, in runInteraction
                                        inner_func, *args, **kwargs
                                      File "/usr/lib/python2.7/site-packages/twisted/python/threadpool.py", line 250, in inContext
                                        result = inContext.theWork()
                                      File "/usr/lib/python2.7/site-packages/twisted/python/threadpool.py", line 266, in <lambda>
                                        inContext.theWork = lambda: context.call(ctx, func, *args, **kw)
                                      File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 122, in callWithContext
                                        return self.currentContext().callWithContext(ctx, func, *args, **kw)
                                      File "/usr/lib/python2.7/site-packages/twisted/python/context.py", line 85, in callWithContext
                                        return func(*args,**kw)
                                      File "/usr/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 307, in _runWithConnection
                                        compat.reraise(excValue, excTraceback)
                                      File "/usr/lib/python2.7/site-packages/twisted/enterprise/adbapi.py", line 298, in _runWithConnection
                                        result = func(conn, *args, **kw)
                                      File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 313, in inner_func
                                        func, *args, **kwargs
                                      File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 242, in _new_transaction
                                        r = func(txn, *args, **kwargs)
                                      File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 485, in _simple_upsert_txn
                                        self.database_engine.lock_table(txn, table)
                                      File "/usr/lib/python2.7/site-packages/synapse/storage/engines/postgres.py", line 64, in lock_table
                                        txn.execute("LOCK TABLE %s in EXCLUSIVE MODE" % (table,))
                                      File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 80, in execute
                                        self._do_execute(self.txn.execute, sql, *args)
                                      File "/usr/lib/python2.7/site-packages/synapse/storage/_base.py", line 110, in _do_execute
                                        sql, *args
                                    InternalError: LOCK TABLE can only be used in transaction blocks
```

... which means that the server ends up not trusting half the other servers in the room, which means it ends up with lots of missing state, which means that it hammers on /event/ for other servers, which means it contributes to https://github.com/matrix-org/synapse/issues/2164

**Comment by richvdh** [comment]
So this does indeed appear to be related to #2135, in that if the 'create index' statement fails, we leave the db connection with 'autocommit=True', which then causes this exception.

**Comment by richvdh** [comment]
Fixed by #2167 I hope
